extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -607,6 +631,47 @@ export class AmpStory360 extends AMP.BaseElement {
       );
   }
 
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  setupAmpVideoRenderer_() {
+    return whenUpgradedToCustomElement(dev().assertElement(this.ampVideoEl_))
+      .then(() => {
+        return this.ampVideoEl_.signals().whenSignal(CommonSignals.LOAD_END);
+      })
+      .then(() => {
+        return (
+          this.ampVideoEl_ && listenOncePromise(this.ampVideoEl_, 'playing')
```

Comment:
This is super tricky and comes with a lot of context, can you tell more about why you want to wait for this event so we can make sure we do it in a safe and not prone to race conditions way?

Reply:
`gl.texImage2D` needs image data from the video element. Without a playing video it receives an invalid value. 

<img width="837" alt="Screen Shot 2020-10-07 at 5 09 52 PM" src="https://user-images.githubusercontent.com/3860311/95388495-5eabea00-08c0-11eb-81af-7f24aac27c0d.png">

This happens in:
https://github.com/ampproject/amphtml/blob/43d58d947cff699915c1b9e9b46bb02b436153ea/third_party/zuho/zuho.js#L159-L164

Reply:
Sounds good thank you.
At this time, it is possible that the `playing` event already fired, or that it will fire but the video element will be replaced by the media pool for a non playing one.

To work around these race conditions I'd suggest we start going through the rAF loop that tries to render the video frame anyway, but add an additional test to make sure the video element has data. I think looking at the video HTMLMediaElement  `readyState` would work, and that `readyState >= 2` looks promising. https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/readyState

Wdyt?

Reply:
I think we need this listener to finish setting up the renderer and to know when to start animating.
Without it that listener we get gl errors from the `setImage` and `renderInitialPosition_` methods.
We also risk the element animating before the video is playing.

What if we do this as an additional check in the rAF loop?

Reply:
You still have potential issues where the video might already be playing, and you won't ever receive any `playing` event.
Can you add a `readyState` check here too, and return a resolved promise if it's already `>= 2`?

Additionally, if `readyState >= 2` works here and in the rAF loop, we should listen for the `loadeddata` event rather than `playing`, which would trigger sooner and make the 360 video visible to users earlier (cf https://html.spec.whatwg.org/multipage/media.html#event-media-loadeddata).
If `loadeddata` doesn't work, we should try `canplay` and `readyState >= 3` (cf the HTMLMediaElement spec)

Reply:
Using the `loadeddata` event instead of `playing` sounds good. 
It will ensure the first render happens as soon as possible.

The `loadeddata` event fires when `readyState === 2` so I don't think we need to check `readyState` here.

Reply:
You do need to check the `readyState`, in case the `loadeddata` event already fired BEFORE you start listening.